unit-tests/source_tree_test.py

```diff
@@ -20,81 +41,73 @@ def test_analyser(self, tmp_path: Path):
         (tmp_path / 'directory' / 'file.foo')\
             .write_text("Second file in directory")
 
-        tracker: List[Path] = []
-
-        class DummyTask(Analyser):
-            def run(self):
-                tracker.append(self._reader.filename)
-
         db = SqliteStateDatabase(tmp_path)
         emap: Dict[str, Union[Type[Task], Type[Command]]] = {
-            '.foo': DummyTask
+            '.foo': DummyAnalyser
         }
         fmap: Dict[Type[Command], List[str]] = {}
         queue = QueueManager(1)
+        queue.run()
         test_unit = ExtensionVisitor(emap, fmap, db, tmp_path, queue)
-        tracker.clear()
+
+        tracker[:] = []
+
         test_unit.visit(tmp_path / 'test.foo')
-        assert tracker == [tmp_path / 'test.foo']
+        queue.check_queue_done()
+        assert list(tracker) == [tmp_path / 'test.foo']
 
         test_unit.visit(tmp_path / 'directory' / 'file.foo')
-        assert tracker == [tmp_path / 'test.foo',
-                           tmp_path / 'directory' / 'file.foo']
+        queue.check_queue_done()
+        assert list(tracker) == [tmp_path / 'test.foo',
+                                 tmp_path / 'directory' / 'file.foo']
+        queue.shutdown()
 
     def test_command(self, tmp_path: Path):
         (tmp_path / 'test.bar').write_text("File the first")
         (tmp_path / 'directory').mkdir()
         (tmp_path / 'directory' / 'test.bar') \
             .write_text("File the second in directory")
 
-        tracker: List[Path] = []
-
-        class DummyCommand(SingleFileCommand):
-            @property
-            def output(self) -> List[Path]:
-                return [Path(tmp_path / 'wiggins')]
-
-            @property
-            def as_list(self) -> List[str]:
-                tracker.append(self._filename)
-                return ['cp', str(self._filename), str(self.output[0])]
-
         db = SqliteStateDatabase(tmp_path)
         emap: Dict[str, Union[Type[Task], Type[Command]]] = {
             '.bar': DummyCommand
         }
         fmap: Dict[Type[Command], List[str]] = {}
         queue = QueueManager(1)
+        queue.run()
         test_unit = ExtensionVisitor(emap, fmap, db, tmp_path, queue)
-        tracker.clear()
+
+        tracker[:] = []
+
         test_unit.visit(tmp_path / 'test.bar')
```

Comment:
Is there a way we can implement the visitor so it can be tested on its own, without the queue stuff? A way in which all the queueing happens in the descent and not the visitor?

Reply:
We could go back to a model where the visitor `yield`s stuff for the descender to deal with (only this time it would be `yield`ing the task that needs to be stuck on the queue)... though we've also got the hash calculator to add which might complicate things

Reply:
In the end I've gone for changing the visitor so it accepts a `Callable` in its constructor that it will call with any `Task` instance it wishes to run - this way the queuing mechanism is pulled up into the application layer (and we can bypass it in the unit tests)